modules/percolator/src/main/java/org/elasticsearch/percolator/PercolatorMatchedSlotSubFetchPhase.java

```diff
@@ -0,0 +1,116 @@
+ */
+class PercolatorMatchedSlotSubFetchPhase implements FetchSubPhase {
+
+    static final String FIELD_NAME = "_percolator_document_slot";
+
+    @Override
+    public void hitsExecute(SearchContext context, SearchHit[] hits) throws IOException {
+        PercolateQuery percolateQuery = locatePercolatorQuery(context.query());
+        if (percolateQuery == null) {
+            return;
+        }
+
+        if (percolateQuery.getDocuments().size() == 1) {
+            for (SearchHit hit : hits) {
+                Map<String, DocumentField> fields = hit.fieldsOrNull();
+                if (fields == null) {
+                    fields = new HashMap<>();
+                    hit.fields(fields);
+                }
+                fields.put(FIELD_NAME, new DocumentField(FIELD_NAME, Collections.singletonList(0)));
+            }
+            return;
+        }
+
+        IndexSearcher percolatorIndexSearcher = percolateQuery.getPercolatorIndexSearcher();
+        Weight weight = percolatorIndexSearcher.createNormalizedWeight(Queries.newNonNestedFilter(), false);
+        Scorer s = weight.scorer(percolatorIndexSearcher.getIndexReader().leaves().get(0));
+        int memoryIndexMaxDoc = percolatorIndexSearcher.getIndexReader().maxDoc();
+        BitSet rootDocs = BitSet.of(s.iterator(), memoryIndexMaxDoc);
+        boolean hasNestedDocs = rootDocs.cardinality() != percolatorIndexSearcher.getIndexReader().numDocs();
+
+        PercolateQuery.QueryStore queryStore = percolateQuery.getQueryStore();
+        List<LeafReaderContext> ctxs = context.searcher().getIndexReader().leaves();
+        for (SearchHit hit : hits) {
+            LeafReaderContext ctx = ctxs.get(ReaderUtil.subIndex(hit.docId(), ctxs));
+            int segmentDocId = hit.docId() - ctx.docBase;
+            Query query = queryStore.getQueries(ctx).apply(segmentDocId);
+
+            TopDocs topDocs = percolatorIndexSearcher.search(query, memoryIndexMaxDoc, new Sort(SortField.FIELD_DOC));
+            Map<String, DocumentField> fields = hit.fieldsOrNull();
+            if (fields == null) {
+                fields = new HashMap<>();
+                hit.fields(fields);
+            }
+            IntStream slots = convertTopDocsToSlots(topDocs, rootDocs, hasNestedDocs);
+            fields.put(FIELD_NAME, new DocumentField(FIELD_NAME, slots.boxed().collect(Collectors.toList())));
+        }
+    }
+
+    static IntStream convertTopDocsToSlots(TopDocs topDocs, BitSet rootDocs, boolean hasNestedDocs) {
+        IntStream stream = Arrays.stream(topDocs.scoreDocs)
+            .mapToInt(scoreDoc -> scoreDoc.doc);
+        if (hasNestedDocs) {
+            stream = stream.map(docId -> mapDocIdToSlot(rootDocs, docId));
+        }
+        return stream;
+    }
+
+    private static int mapDocIdToSlot(BitSet rootDocs, int docId) {
+        int slot = -1;
+        for (; docId > 0; docId = rootDocs.prevSetBit(docId - 1)) {
+            slot++;
+        }
+        return slot;
```

Comment:
This will need to be repeated for every hit and runs in linear time with the number of percolated documents. Maybe we could decrease the cost of this operation by building an int[] that stores the doc ids of root documents up-front and then using Arrays.binarySearch to find the slot?

Reply:
Good idea, I'll try that.